shared/dataflow/codeql/dataflow/test/ProvenancePathGraph.qll

```diff
@@ -0,0 +1,82 @@
+/**
+ * Provides a module for renumbering MaD IDs in data flow path explanations in
+ * order to produce more stable test output.
+ *
+ * In addition to the `PathGraph`, a `query predicate models` is provided to
+ * list the contents of the referenced MaD rows.
+ */
+signature predicate interpretModelForTestSig(QlBuiltins::ExtensionId madId, string model);
+
+signature class PathNodeSig {
+  string toString();
+}
+
+signature module PathGraphSig<PathNodeSig PathNode> {
+  /** Holds if `(a,b)` is an edge in the graph of data flow path explanations. */
+  predicate edges(PathNode a, PathNode b, string key, string val);
+
+  /** Holds if `n` is a node in the graph of data flow path explanations. */
+  predicate nodes(PathNode n, string key, string val);
+
+  /**
+   * Holds if `(arg, par, ret, out)` forms a subpath-tuple, that is, flow through
+   * a subpath between `par` and `ret` with the connecting edges `arg -> par` and
+   * `ret -> out` is summarized as the edge `arg -> out`.
+   */
+  predicate subpaths(PathNode arg, PathNode par, PathNode ret, PathNode out);
+}
+
+module ShowProvenance<
+  interpretModelForTestSig/2 interpretModelForTest, PathNodeSig PathNode,
+  PathGraphSig<PathNode> PathGraph>
+{
+  private predicate madIds(string madId) {
+    exists(string model |
+      PathGraph::edges(_, _, _, model) and
+      model.regexpFind("(?<=MaD:)[0-9]*", _, _) = madId
+    )
+  }
+
+  private predicate rankedMadIds(string madId, int r) {
+    madId = rank[r](string madId0 | madIds(madId0) | madId0 order by madId0.toInt())
+  }
+
+  query predicate models(int r, string model) {
+    exists(QlBuiltins::ExtensionId madId |
+      rankedMadIds(madId.toString(), r) and interpretModelForTest(madId, model)
+    )
+  }
+
+  private predicate translateModelsPart(string model1, string model2, int i) {
+    PathGraph::edges(_, _, _, model1) and
+    exists(string s | model1.splitAt("MaD:", i) = s |
+      model2 = s and i = 0
+      or
+      exists(string part, string madId, string rest, int r |
+        translateModelsPart(model1, part, i - 1) and
+        madId = s.regexpCapture("([0-9]*)(.*)", 1) and
+        rest = s.regexpCapture("([0-9]*)(.*)", 2) and
+        rankedMadIds(madId, r) and
+        model2 = part + "MaD:" + r + rest
+      )
+    )
+  }
+
+  private predicate translateModels(string model1, string model2) {
+    exists(int i |
+      translateModelsPart(model1, model2, i) and
+      not translateModelsPart(model1, _, i + 1)
+    )
+  }
+
+  query predicate edges(PathNode a, PathNode b, string key, string val) {
```

Comment:
Is there a reason for not printing the model itself instead of its "translated" id? Even though it is less likely, we would still need to update tests in case models suddenly swap order (not sure how stable the QlBuiltsins::ExtensionId is)? Furthermore, it might be easier to "identify" the model used by the edge from MaD format instead of an integer.

Reply:
Also, the translated ids could change in case a new edge is added that uses a model, which changes (pushes) the ranking (compared to the last time the test was run). This could mean that a large "unrelated" part of an expected file needs to be updated (which might cause confusion)

Reply:
I don't believe order swapping to be very likely, although @dbartol knows the full details for a more precise answer to that.
I do realise that there's shifting going on when a test changes the set of models that it's referring to, but I thought it was nicer to have the models on the side in order to have a narrower edges relation (narrow in the sense that the row in the `.expected` file fits within the width of a reasonably sized editor window). And balancing these two things, I think I'd prefer narrower edges over no shifting.

Reply:
That is a valid argument. I am just worried that "interesting" updates to the expected file could drown in noise. Also, it introduces added complexity to the expected files that one needs to know of the possibility of shifting.
It is ok with me, if we stick with the current solution, if we are willing to change it in case we frequently run into the  problem above.